File: src/context/GameContext.java
package context;

import java.awt.*;
import java.awt.image.BufferedImage;
import java.util.ArrayList;

public class GameContext {
        // Current dimensions (updated on resize)
    public static int BOARD_WIDTH = DimensionalContext.BASE_WIDTH;
    public static int BOARD_HEIGHT = DimensionalContext.BASE_HEIGHT;
    public static int PADDLE_WIDTH = 20;
    public static int PADDLE_HEIGHT = 80;
    public static int BALL_SIZE = 20;
    public static int MENU_PADDLE_WIDTH = 40;

    // Background selection variables
    public static int selectedBackground = 0; // Default background
    public static ArrayList<String> backgroundNames = new ArrayList<>(); // Dynamic background names
    public static ArrayList<Image> backgroundImages = new ArrayList<>(); // Loaded background images
    public static ArrayList<String> backgroundFilenames = new ArrayList<>(); // Original filenames for lazy loading
    public static int selectedBackgroundOption = 0; // Currently selected in background menu

    // Lazy loading flags (web mode optimization)
    public static boolean allBackgroundsLoaded = false; // Track if all backgrounds are loaded
    public static boolean allPaddleThemesLoaded = false; // Track if all paddle themes are loaded

    // Paddle selection variables
    // Separate paddle themes for blue (left) and red (right) paddles
    public static ArrayList<String> bluePaddleThemeNames = new ArrayList<>(); // Blue paddle themes
    public static ArrayList<BufferedImage> bluePaddleThemeImages = new ArrayList<>(); // Blue paddle images
    public static ArrayList<String> bluePaddleThemeFilenames = new ArrayList<>(); // Original filenames for lazy loading
    public static ArrayList<String> redPaddleThemeNames = new ArrayList<>(); // Red paddle themes
    public static ArrayList<BufferedImage> redPaddleThemeImages = new ArrayList<>(); // Red paddle images
    public static ArrayList<String> redPaddleThemeFilenames = new ArrayList<>(); // Original filenames for lazy loading

    // Legacy arrays for compatibility (will use blue themes for now)
    public static ArrayList<String> paddleThemeNames = new ArrayList<>(); // Available paddle themes
    public static ArrayList<BufferedImage> paddleThemeImages = new ArrayList<>(); // Loaded paddle images
    public static int selectedPaddleTheme = 0; // Currently selected left paddle theme
    public static int selectedRightPaddleTheme = 0; // Currently selected right paddle theme
    public static int previewPaddleY = 300; // Y position of preview paddle in selection screen

    // Game loop constants
    public static final int LOGIC_FPS = 60;
    public static final long LOGIC_TIME_STEP = 1000000000L / LOGIC_FPS; // nanoseconds
    public static volatile boolean gameRunning = false;
    public static Thread gameLoopThread;


    // Scale factors
    public static double scaleX = 1.0;
    public static double scaleY = 1.0;

}

File: src-js/context/GameContext.js
/**
 * Game Context
 * Central state management for game dimensions, assets, and runtime variables
 */

import { BASE_WIDTH, BASE_HEIGHT } from './DimensionalContext.js';

// Current dimensions (updated on resize)
export let BOARD_WIDTH = BASE_WIDTH;
export let BOARD_HEIGHT = BASE_HEIGHT;
export let PADDLE_WIDTH = 20;
export let PADDLE_HEIGHT = 80;
export let BALL_SIZE = 20;
export let MENU_PADDLE_WIDTH = 40;

// Functions to update dimensions
export function setBoardDimensions(width, height) {
    BOARD_WIDTH = width;
    BOARD_HEIGHT = height;
}

export function setPaddleDimensions(width, height) {
    PADDLE_WIDTH = width;
    PADDLE_HEIGHT = height;
}

export function setBallSize(size) {
    BALL_SIZE = size;
}

export function setMenuPaddleWidth(width) {
    MENU_PADDLE_WIDTH = width;
}

// Background selection variables
export let selectedBackground = 0; // Default background
export let backgroundNames = []; // Dynamic background names
export let backgroundImages = []; // Loaded background images
export let backgroundFilenames = []; // Original filenames for lazy loading
export let selectedBackgroundOption = 0; // Currently selected in background menu

// Lazy loading flags (web mode optimization)
export let allBackgroundsLoaded = false; // Track if all backgrounds are loaded
export let allPaddleThemesLoaded = false; // Track if all paddle themes are loaded

// Paddle selection variables
// Separate paddle themes for blue (left) and red (right) paddles
export let bluePaddleThemeNames = []; // Blue paddle themes
export let bluePaddleThemeImages = []; // Blue paddle images
export let bluePaddleThemeFilenames = []; // Original filenames for lazy loading
export let redPaddleThemeNames = []; // Red paddle themes
export let redPaddleThemeImages = []; // Red paddle images
export let redPaddleThemeFilenames = []; // Original filenames for lazy loading

// Legacy arrays for compatibility (will use blue themes for now)
export let paddleThemeNames = []; // Available paddle themes
export let paddleThemeImages = []; // Loaded paddle images
export let selectedPaddleTheme = 0; // Currently selected left paddle theme
export let selectedRightPaddleTheme = 0; // Currently selected right paddle theme
export let previewPaddleY = 300; // Y position of preview paddle in selection screen

// Game loop constants
export const LOGIC_FPS = 60;
export const LOGIC_TIME_STEP = 1000000000 / LOGIC_FPS; // nanoseconds (will be in milliseconds for JS)
export let gameRunning = false;
export let gameLoopThread = null;

// Scale factors
export let scaleX = 1.0;
export let scaleY = 1.0;

// Setters for modifying exported values
export function setSelectedBackground(value) {
    selectedBackground = value;
}

export function setSelectedBackgroundOption(value) {
    selectedBackgroundOption = value;
}

export function setAllBackgroundsLoaded(value) {
    allBackgroundsLoaded = value;
}

export function setAllPaddleThemesLoaded(value) {
    allPaddleThemesLoaded = value;
}

export function setSelectedPaddleTheme(value) {
    selectedPaddleTheme = value;
}

export function setSelectedRightPaddleTheme(value) {
    selectedRightPaddleTheme = value;
}

export function setPreviewPaddleY(value) {
    previewPaddleY = value;
}

export function setGameRunning(value) {
    gameRunning = value;
}

export function setGameLoopThread(value) {
    gameLoopThread = value;
}

export function setScaleFactors(x, y) {
    scaleX = x;
    scaleY = y;
}
